pkg/controller/sync/updater.go

```diff
@@ -0,0 +1,222 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sync
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+)
+
+type FederatedUpdater interface {
+	NoChanges() bool
+	Wait() (map[string]string, bool)
+
+	Create(clusterName string)
+	Update(clusterName string, clusterObj *unstructured.Unstructured)
+	Delete(clusterName string)
+}
+
+type updaterResult struct {
+	clusterName string
+	version     string
+	ok          bool
+}
+
+type federatedUpdaterImpl struct {
+	sync.RWMutex
+
+	fedView util.FederationView
+
+	fedResource FederatedResource
+
+	resultChan     chan updaterResult
+	operationCount int
+
+	timeout time.Duration
+}
+
+func NewFederatedUpdater(fedView util.FederationView, fedResource FederatedResource) FederatedUpdater {
+	return &federatedUpdaterImpl{
+		fedView:     fedView,
+		fedResource: fedResource,
+		resultChan:  make(chan updaterResult),
+		timeout:     30 * time.Second, // TODO(marun) Make this configurable
+	}
+}
+
+func (u *federatedUpdaterImpl) NoChanges() bool {
+	return u.operationCount == 0
+}
+
+func (u *federatedUpdaterImpl) Wait() (map[string]string, bool) {
+	versions := make(map[string]string)
+	ok := true
+
+	timedOut := false
+	start := time.Now()
+	for i := 0; i < u.operationCount; i++ {
+		now := time.Now()
+		if !now.Before(start.Add(u.timeout)) {
+			timedOut = true
+			break
+		}
+		select {
+		case result := <-u.resultChan:
+			if !result.ok {
+				ok = false
+				break
+			}
+			if len(result.version) > 0 {
+				versions[result.clusterName] = result.version
+			}
+		case <-time.After(start.Add(u.timeout).Sub(now)):
+			timedOut = true
+			break
+		}
+	}
+	if timedOut {
+		ok = false
+		u.fedResource.RecordError("OperationTimeoutError", errors.Errorf("Failed to finish all operations in %v", u.timeout))
+	}
+
+	return versions, ok
+}
+
+func (u *federatedUpdaterImpl) Create(clusterName string) {
+	u.operationCount += 1
+	const op = "create"
+	go u.clusterOperation(clusterName, op, func(client util.ResourceClient) (string, error) {
+		u.recordEvent(clusterName, op, "Creating")
+
+		obj, err := u.fedResource.ObjectForCluster(clusterName)
+		if err != nil {
+			return "", err
+		}
+		createdObj, err := client.Resources(obj.GetNamespace()).Create(obj, metav1.CreateOptions{})
```

Comment:
(query) Is there a possibility that an client operation gets stuck and never return? If that happens there is a possibility that the go routine would be leaked and many such incidences would keep increasing memory.

Reply:
The behavior here is the same as the old updater, and any request should eventually timeout (it just might not be within 30s).  As per a TODO in this file, I am intending to switch to the generic client to take advantage of context cancellation, which should ensure immediate cleanup rather than timeout.

Reply:
Ack